web/src/enterprise/campaigns/detail/changesets/CampaignChangesets.tsx

```diff
@@ -7,92 +7,114 @@ import { FilteredConnection, FilteredConnectionQueryArgs, Connection } from '../
+import { repeatWhen, delay } from 'rxjs/operators'
 
 interface Props extends ThemeProps {
-    queryChangesetsConnection: (
-        args: FilteredConnectionQueryArgs
-    ) => Observable<Connection<GQL.IExternalChangeset | GQL.IChangesetPlan>>
+    campaign: Pick<GQL.ICampaignPlan, '__typename' | 'id'> | Pick<GQL.ICampaign, '__typename' | 'id' | 'closedAt'>
     history: H.History
     location: H.Location
     campaignUpdates: Subject<void>
     changesetUpdates: Subject<void>
-    /** Shows the publish button for ChangesetPlans */
-    enablePublishing: ChangesetNodeProps['enablePublishing']
+
+    /** For testing only. */
+    _queryChangesets?: (
+        campaignID: GQL.ID,
+        args: FilteredConnectionQueryArgs
+    ) => Observable<Connection<GQL.IExternalChangeset | GQL.IChangesetPlan>>
 }
 
 /**
  * A list of a campaign's or campaign preview's changesets.
  */
 export const CampaignChangesets: React.FunctionComponent<Props> = ({
-    queryChangesetsConnection: _queryChangesetsConnection,
+    campaign,
     history,
     location,
     isLightTheme,
     changesetUpdates,
     campaignUpdates,
-    enablePublishing,
+    _queryChangesets = queryChangesets,
 }) => {
     const [state, setState] = useState<GQL.ChangesetState | undefined>()
     const [reviewState, setReviewState] = useState<GQL.ChangesetReviewState | undefined>()
     const [checkState, setCheckState] = useState<GQL.ChangesetCheckState | undefined>()
+
     const queryChangesetsConnection = useCallback(
-        args => _queryChangesetsConnection({ ...args, state, reviewState, checkState }),
-        [_queryChangesetsConnection, state, reviewState, checkState]
+        (args: FilteredConnectionQueryArgs) => {
+            const queryObservable: Observable<
+                GQL.IChangesetPlanConnection | Connection<GQL.IExternalChangeset | GQL.IChangesetPlan>
+            > =
+                campaign.__typename === 'CampaignPlan'
+                    ? queryChangesetPlans(campaign.id, args)
+                    : _queryChangesets(campaign.id, { ...args, state, reviewState, checkState })
+            return queryObservable.pipe(repeatWhen(obs => obs.pipe(delay(5000))))
+        },
+        [campaign.id, campaign.__typename, state, reviewState, checkState, _queryChangesets]
+    )
+
+    const changesetFiltersRow = (
+        <div className="form-inline mb-0 mt-2">
+            <label htmlFor="changeset-state-filter">State</label>
+            <select
+                className="form-control mx-2"
+                value={state}
+                onChange={e => setState((e.target.value || undefined) as GQL.ChangesetState | undefined)}
+                id="changeset-state-filter"
+            >
+                <option value="">All</option>
+                {Object.values(GQL.ChangesetState).map(state => (
+                    <option value={state} key={state}>
+                        {upperFirst(state.replace(/_/g, ' ').toLocaleLowerCase())}
```

Comment:
You can use lodash for this to avoid the duplication
```suggestion
                        {upperFirst(lowerCase(state))}
```